### PR TITLE
Flink: Add RowConverter for Iceberg Source

### PR DESCRIPTION
**Pull request by abharath9**
Currently we can't create views on top of IcebergSource DataStreams directly. We need to convert the RowData to Row explicitly using map function. I thought creating a RowConverter to convert RowData to Row and return IcebergSource<Row> would be a good idea. This approach enables the use of the Iceberg schema to create a DataStream<Row> and further create a view on top of it. What are your thoughts?

### CODE REVIEW

**Comment by abharath9** [comment]
@stevenzwu  can i get a review for this

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/source/reader/RowConverter.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import org.apache.flink.table.data.conversion.DataStructureConverter;
+import org.apache.flink.table.data.conversion.DataStructureConverters;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.RowType;
+import org.apache.flink.table.types.utils.TypeConversions;
+import org.apache.flink.types.Row;
+import org.apache.iceberg.flink.FlinkSchemaUtil;
+import org.jetbrains.annotations.NotNull;
+
+public class RowConverter implements RowDataConverter<Row> {
```
Comment (stevenzwu) [inline flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/source/reader/RowConverter.java:33]: code probably can be simplified. see example from `RowDataToRowMapper`

```
  @Override
  public void open(Configuration parameters) throws Exception {
    this.converter =
        DataStructureConverters.getConverter(TypeConversions.fromLogicalToDataType(rowType));
  }

  @Override
  public Row map(RowData value) throws Exception {
    return (Row) converter.toExternal(value);
  }
}
```

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedRow.java [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.types.TypeUtil;
+import org.junit.jupiter.api.TestTemplate;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.junit.jupiter.api.io.TempDir;
+
+@ExtendWith(ParameterizedTestExtension.class)
+public class TestIcebergSourceBoundedRow {
```
Comment (stevenzwu) [inline flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedRow.java:65]: we can probably reduce the code duplication with `TestIcebergSourceBoundedGenericRecord` with either extracting some common code to a base class or use parameterization.

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/source/reader/RowConverter.java (hunk 1 above)
Comment (abharath9) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/source/reader/RowConverter.java:33]: thanks for the feedback. Do you mean change this to RowData To Row mapper?

Comment (stevenzwu) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/source/reader/RowConverter.java:33]: I meant refer to the `RowDataToRowMapper` for simpler implementation of this converter.

Comment (abharath9) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/source/reader/RowConverter.java:32]: @stevenzwu simplified the code. Can you take a look? If it looks good then i can make changes to the test cases.

Currently, there’s no utility function to extract RowTypeInfo from RowType, so I have to use the RowTypeInfo class to create it. Not sure if this is the correct way but let me know if there's a better way. Thanks

Comment (stevenzwu) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/source/reader/RowConverter.java:32]: LGTM

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedRow.java (hunk 2 above)
Comment (abharath9) [reply flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedRow.java]: @stevenzwu Refactored the test cases. Please take a look.

Comment (abharath9) [reply flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedRow.java]: @stevenzwu can you take a look at the pr. If everything looks good we can merge the pr. Thanks

**Comment by abharath9** [comment]
@stevenzwu Hope you are doing well. The pr has been open for a while. If everything looks good can you approve the pr. Thanks

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedRow.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.types.TypeUtil;
+import org.junit.jupiter.api.TestTemplate;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+@ExtendWith(ParameterizedTestExtension.class)
+public class TestIcebergSourceBoundedRow extends TestIcebergSourceBoundedConverterBase<Row> {
+
+  @TestTemplate
+  public void testUnpartitionedTable() throws Exception {
```
Comment (stevenzwu) [inline flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedRow.java:48]: can these test methods be moved to the base class?

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedGenericRecord.java [hunk 4]
```diff
@@ -18,112 +18,77 @@
...
 
   @Parameters(name = "format={0}, parallelism = {1}, useConverter = {2}")
   public static Object[][] parameters() {
-    return new Object[][] {
-      {FileFormat.AVRO, 2, true},
-      {FileFormat.PARQUET, 2, true},
-      {FileFormat.PARQUET, 2, false},
-      {FileFormat.ORC, 2, true}
+    return new Object[][]{
+        {FileFormat.AVRO, 2, true},
```
Comment (stevenzwu) [inline flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedGenericRecord.java:60]: there seems to have check style error in the CI build due to this change

Comment (abharath9) [reply flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedGenericRecord.java]: only these lines are automatically indenting with 4 spaces every time I reformat the file. I changed them and they should indent with 2 spaces now.

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedRow.java (hunk 3 above)
Comment (abharath9) [reply flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedRow.java]: Good idea. I refactored the code. Please check

**Comment by abharath9** [comment]
@stevenzwu Seems like format violations. I will take a look. Is there a way to run all these checks after committing the changes?

**Comment by abharath9** [comment]
@stevenzwu formatting violations should be fixed now. Whenever you get a chance can you run the build?

**Review by stevenzwu — APPROVED** [review]

**Comment by stevenzwu** [comment]
@abharath9 thanks for the contribution. 

can you also create clean back port to 1.18 and 1.19

**Comment by abharath9** [comment]
> @abharath9 thanks for the contribution.
> 
> can you also create clean back port to 1.18 and 1.19

@stevenzwu can you also review this pr https://github.com/apache/iceberg/pull/11838

**Pull request merged**